server/service/service.go

```diff
@@ -269,29 +275,204 @@ func (s *Service) ReportStatus(ctx context.Context, req *bpb.ReportStatusRequest
 	return &bpb.EmptyResponse{}, s.em.SetStatus(ctx, req)
 }
 
+// Function to check if a valid client cert was presented
+func hasClientCert(ctx context.Context) bool {
+	p, ok := peer.FromContext(ctx)
+	if !ok {
+		return false
+	}
+	tlsInfo, ok := p.AuthInfo.(credentials.TLSInfo)
+	if !ok {
+		return false
+	}
+	return len(tlsInfo.State.PeerCertificates) > 0
+}
+
+// verifyNonceSignature verifies the signature over the nonce.
+func verifyNonceSignature(pubKey crypto.PublicKey, nonce, signature []byte) error {
+	rsaPubKey, ok := pubKey.(*rsa.PublicKey)
+	if !ok {
+		return log.Errorf("public key is not RSA")
+	}
+	hash := sha256.Sum256(nonce)
+	err := rsa.VerifyPSS(rsaPubKey, crypto.SHA256, hash[:], signature, &rsa.PSSOptions{
+		SaltLength: rsa.PSSSaltLengthAuto,
+		Hash:       crypto.SHA256,
+	})
+	if err != nil {
+		return log.Errorf("signature verification failed: %w", err)
+	}
+	return nil
+}
+
+const (
+	stateInitial = iota
+	stateNonceSent
+	stateAttested
+)
+
 func (s *Service) BootstrapStream(stream bpb.Bootstrap_BootstrapStreamServer) error {
+	ctx := stream.Context()
+	clientHasCert := hasClientCert(ctx)
+
+	var deviceID string
+	var chassis *epb.Chassis
+	var activeControlCard string
+	currentState := stateInitial
+
+	if clientHasCert {
+		p, _ := peer.FromContext(ctx)
+		tlsInfo := p.AuthInfo.(credentials.TLSInfo)
+		clientCert := tlsInfo.State.PeerCertificates[0]
+		deviceID = clientCert.Subject.SerialNumber
+		if deviceID == "" {
+			deviceID = clientCert.Subject.CommonName
+		}
+		if deviceID == "" {
+			return status.Errorf(codes.Unauthenticated, "client certificate missing unique identifier")
+		}
+		log.Infof("Stream opened by device with IDevID: %s", deviceID)
+	} else {
+		log.Info("Stream opened by device without client certificate")
+	}
+
 	for {
 		in, err := stream.Recv()
-		if err == io.EOF { // Client is done.  Success.
-			return nil // The RPC library will call stream.CloseSend for us.
+		if err == io.EOF { // Client is done.
+			log.Infof("Stream closed by client: %s", deviceID)
+			s.nonces.Delete(deviceID) // Clean up session
+			return nil
 		}
 		if err != nil {
-			return err // The RPC library will call stream.Abort(err) for us.
+			log.Errorf("Stream error for device %s: %v", deviceID, err)
+			s.nonces.Delete(deviceID) // Clean up session
+			return err
 		}
+
 		switch req := in.Type.(type) {
 		case *bpb.BootstrapStreamRequest_BootstrapRequest:
-			// TODO: process request
+			if currentState != stateInitial {
+				return status.Errorf(codes.FailedPrecondition, "received BootstrapRequest in unexpected state: %v", currentState)
+			}
+			bootstrapReq := req.BootstrapRequest
+			identity := bootstrapReq.GetIdentity()
+
+			if clientHasCert {
+				// --- TPM 2.0 with IDevID Flow ---
```

Comment:
Similar to above, the IDevID flow in BootstrapStream doesn't rely on any client cert. You need to check Identity() in the Bootstrap request instead